frontend/apps/reader/modules/readerpagemap.lua

```diff
@@ -39,20 +42,47 @@ function ReaderPageMap:init()
     self.ui:registerPostInitCallback(function()
         self:_postInit()
     end)
+    self.ui.menu:registerToMainMenu(self)
 end
 
 function ReaderPageMap:_postInit()
     self.initialized = true
-    if self.ui.document.info.has_pages then
-        return
+    -- chars_per_synthetic_page is saved to cr3 cache by crengine on first building of synthetic pagemap
+    -- after that synthetic pagemap is built on doc rendering without asking doc settings
+    -- different combinations of doc settings and cached chars_per_synthetic_page must be handled
+    local cached_chars_per_synthetic_page = self.ui.document:getSyntheticPageMapCharsPerPage()
+    if self.ui.document.is_new then
+        if cached_chars_per_synthetic_page > 0 then -- honor cr3 cache
+            self.chars_per_synthetic_page = cached_chars_per_synthetic_page
+            self.ui.doc_settings:saveSetting("pagemap_chars_per_synthetic_page", cached_chars_per_synthetic_page)
+        else -- honor default settings
+            local chars_per_synthetic_page = G_reader_settings:readSetting("pagemap_chars_per_synthetic_page")
+            if chars_per_synthetic_page and
+                    (not self.ui.document:hasPageMap() or G_reader_settings:isTrue("pagemap_synthetic_overrides")) then
```

Comment:
I find it strange you have not used at all hasDocumentProvided or isDocumeentProvided.
Shouldn't you use hasDocumentProvided() here to make it more explicite, than just hasPageMap?